wallet, bench: Move commonly used functions to their own file and fix a bug

achow101 opened the pull request:
I have a few PRs and branches that use these two commits, probably makes sense to split them into a separate PR to be merged sooner.

The first commit contains some things that end up being commonly used in new wallet benchmarks. These are moved into `wallet_common.{h/cpp}`.

The second commit contains a bugfix for the wallet_balance benchmark where it calls `LoadWallet` in the wrong place. It's unnecessary to call that function in this benchmark. Although this does not cause any issues currently, it ends up causing issues in some PRs and branches that I'm working on.

achow101 committed:
bench: Remove incorrect LoadWallet call in WalletBalance

The WalletBalance benchmarks would incorrectly call LoadWallet after the
wallet has been setup. LoadWallet expects to be the first thing that is
called and for the CWallet to be in a fresh state. When it is not, it
results in bogus pointers which can cause segfaults during this
benchmark.

furszy commented on src/bench/util/wallet_common.cpp:
```diff
@@ -0,0 +1,26 @@
+    bilingual_str error;
+    std::vector<bilingual_str> warnings;
+    auto wallet = CWallet::Create(context, "", std::move(database), create_flags, error, warnings);
+    NotifyWalletLoaded(context, wallet);
+    if (context.chain) {
+        wallet->postInitProcess();
+    }
```
little topic:
it seems odd to re-use this two calls in other bench contexts.

`NotifyWalletLoaded` is looping over the wallet load callbacks, and `postInitProcess` syncs the wallet with the mempool, so guess that both will be no-op in most cases?.

achow101 replied:
Yes, I think they'll be no-op for the most part.

furszy commented on src/bench/util/wallet_common.h:
```diff
@@ -0,0 +1,21 @@
+#include <wallet/context.h>
+#include <wallet/wallet.h>
+
+using wallet::CWallet;
+using wallet::DatabaseOptions;
+using wallet::WalletContext;
+using wallet::WalletDatabase;
```
tiny nit: could wrap the functions around a "namespace wallet {}" instead.

achow101 replied:
Dropped `wallet_common.h`, but still ended up doing this in `wallet_loading` and `wallet_balance`

furszy reviewed (commented):
Code ACK, small last topic:

Seems that `BenchLoadWallet` and `BenchUnloadWallet` are similar to `TestLoadWallet` and `TestUnloadWallet` from `wallet_tests.cpp` . The only diff seems to be in the db.

Maybe instead of creating a new file, we could move the bench functions to `wallet/test/util.h` and `wallet/test/util.cpp` so they are used equally for tests and benchmarks?

achow101 committed:
tests, bench: Consolidate {Test,Bench}Un/LoadWallet helper

The wallet tests and benchmarks both had helper functions for loading
and unloading the wallet for the test that were almost identical.
These functions are consolidated and reused.

achow101 committed:
tests: Move ADDRESS_BCRT1_UNSPENDABLE to wallet/test/util.h

This static address is usable for other wallet tests and benchmarks, so
make it available to them.

achow101 commented:
> Maybe instead of creating a new file, we could move the bench functions to `wallet/test/util.h` and `wallet/test/util.cpp` so they are used equally for tests and benchmarks?

Good point, made that change.

It required a few changes to the `CreateWallet` test since that wasn't expecting `postInitProcess` to be run, but should be straightforward to review.

furszy commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -885,7 +866,9 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
-    BOOST_CHECK_EQUAL(addtx_count, 2);
+    // Since mempool transactions are requested at the end of loading, there will
+    // be 2 additional AddToWallet calls, one from the previous test, and a duplicate for mempool_tx
+    BOOST_CHECK_EQUAL(addtx_count, 2 + 2);
```
Not for this PR as it touches the wallet internals but the duplicated `AddToWallet` calls shouldn't be needed if the wallet wtx state is equal to the one in the event.

furszy reviewed (approved):
ACK 7379a54

fanquake commented:
@furszy looks like you ACK'd the wrong commit here.

furszy commented:
oops, fixed. That was the one from the comment. Thanks @fanquake.
